perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1023,56 +1104,70 @@ protected boolean hasCandidateExistingMember(Candidate candidate, RichMember ric
 			List<RichMember> membersToRemove = new ArrayList<>();
-
-			//get all actual members of group
+			// Get all actual members of group
 			List<RichMember> actualGroupMembers = getPerunBl().getGroupsManagerBl().getGroupRichMembers(sess, group);
+			// Subjects gained from external source
+			List<Map<String, String>> subjects = new ArrayList<>();
 
+			String status;
 			if(lightweightSynchronization) {
-				categorizeMembersForLightweightSynchronization(sess, group, source, membersSource, actualGroupMembers, candidatesToAdd, membersToRemove, skippedMembers);
+				status = GroupsManager.GROUP_SYNC_STATUS_LIGHTWEIGHT;
+				// Get subjects from extSource
+				status = getSubjectsFromExtSource(sess, source, group, status, subjects);
+
+				if (status.equals(GroupsManager.GROUP_SYNC_STATUS_FULL)) {
+					categorizeMembersForLightweightSynchronization(sess, source, membersSource, actualGroupMembers,
+							candidatesToAdd, membersToRemove, skippedMembers, subjects);
+				}
 			} else {
-				//Get subjects from extSource
-				List<Map<String, String>> subjects = getSubjectsFromExtSource(sess, source, group);
-				//Convert subjects to candidates
-				List<Candidate> candidates = convertSubjectsToCandidates(sess, subjects, membersSource, source, skippedMembers);
+				status = GroupsManager.GROUP_SYNC_STATUS_FULL;
+				// Get subjects from extSource
+				status = getSubjectsFromExtSource(sess, source, group, status, subjects);
 
-				categorizeMembersForSynchronization(sess, actualGroupMembers, candidates, candidatesToAdd, membersToUpdate, membersToRemove);
+				// Categorizes members for synchronization
+				categorizeMembersForSynchronization(sess, status, source, membersSource, actualGroupMembers, group,
+						candidatesToAdd, membersToUpdate, membersToRemove, skippedMembers, subjects, overwriteUserAttributesList);
 			}
 
-			//Update members already presented in group
-			updateExistingMembersWhileSynchronization(sess, group, membersToUpdate, overwriteUserAttributesList, mergeMemberAttributesList);
+			log.debug("Candidates to add: {} for group: {}", candidatesToAdd.size(), group);
+			log.debug("Members to update: {} for group: {}", membersToUpdate.size(), group);
+			log.debug("Members to remove: {} for group: {}", membersToRemove.size(), group);
 
-			//Add not presented candidates to group
+			// Update members already presented in group
+			updateExistingMembersWhileSynchronization(sess, group, membersToUpdate, overwriteUserAttributesList, mergeMemberAttributesList);
+			// Add not presented candidates to group
 			addMissingMembersWhileSynchronization(sess, group, candidatesToAdd, overwriteUserAttributesList, mergeMemberAttributesList, skippedMembers);
-
-			//Remove presented members in group who are not presented in synchronized ExtSource
-			removeFormerMembersWhileSynchronization(sess, group, membersToRemove);
+			// If status is full, remove members which are not already members of group
```

Comment:
Because synchronization status "full" means, that we have all subjects from external source. If status is "modified", we only have modified members of group, so we can't find out, which members we should remove. Plus there should be no one to remove, so its unnecessary to call this method.